Add support for MXRT600

mmahadevan108 opened the pull request:
This is basic support. The code can run sample/hello_world and sample/basic/blinky applications.

mmahadevan108 committed:
MXRT600: Add support for MXRT600 EVK board

Add support for MXRT600 EVK board

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: gpio: mcux_lpc: Update for RT600

- Use common clock names to avoid build error.
- Use GPIO_PortInit() to do reset/clock initialization. Only clock
  enable is not adequate for RT600.
- Update to support differences between IOCON and IOPCTL modules

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
MXRT600: soc folder

Add a soc folder for rt6xx family

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
MXRT600: DTS file

Add DTS file to support secure and non-secure access

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
pinmux: mxrt6xx: Update to support IOPCTL

Add changes for differences between IOPCTL and IOCON modules

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
modules: hal_nxp: Add support for MXRT600

Add support for MXRT600

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
cc @MaureenHelm

MaureenHelm commented on boards/arm/mimxrt685_evk/board.cmake:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+board_set_debugger_ifnset(jlink)
+board_set_flasher_ifnset(jlink)
```
Remove these lines. Including `jlink.board.cmake` below takes care of this

mmahadevan108 replied:
Done

MaureenHelm commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.yaml:
```diff
@@ -0,0 +1,20 @@
+  - zephyr
+  - gnuarmemb
+  - xtools
+ram: 4608
```
This is a weird number. Is it right?

mmahadevan108 replied:
RT600 has 4.5MB SRAM, so it’s 4608KB.

MaureenHelm commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -0,0 +1,131 @@
+	return 0;
+}
+
+SYS_INIT(mimxrt685_evk_pinmux_init, PRE_KERNEL_1, 45);
```
Why is the priority hard-coded to 45? It should be `CONFIG_PINMUX_INIT_PRIORITY`

mmahadevan108 replied:
Done, thank you for spotting this.

MaureenHelm commented on soc/arm/nxp_imx/Kconfig:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017, NXP
+# Copyright (c) 2020, NXP
```
```suggestion
# Copyright (c) 2017,2020, NXP
```

mmahadevan108 replied:
Done

MaureenHelm commented on soc/arm/nxp_imx/rt6xx/quick_access.ld:
```diff
@@ -0,0 +1,9 @@
+
+. = ALIGN(4);
+KEEP(*(CodeQuickAccess))
+KEEP(*(DataQuickAccess))
```
Does the nxp hal have symbols in these sections?

mmahadevan108 replied:
QuickAccess sections is used in SDK infrastructure when we need certain sections of code to be available in RAM instead of FLEXSPI.

MaureenHelm reviewed (changes requested):
Thanks @mmahadevan108 !

Please add a board document. See `boards/arm/lpcxpresso55s69/doc/index.rst` and `boards/arm/mimxrt1050_evk/doc/index.rst` for example.

Please also fix the errors flagged in the failed compliance check. See https://github.com/zephyrproject-rtos/zephyr/pull/26785/checks?check_run_id=856037114

Moving the nxp hal update commit into this PR will help fix the failed buildkite check.

mmahadevan108 commented:
> Thanks @mmahadevan108 !
> 
> Please add a board document. See `boards/arm/lpcxpresso55s69/doc/index.rst` and `boards/arm/mimxrt1050_evk/doc/index.rst` for example.
> 
> Please also fix the errors flagged in the failed compliance check. See https://github.com/zephyrproject-rtos/zephyr/pull/26785/checks?check_run_id=856037114
> 
> Moving the nxp hal update commit into this PR will help fix the failed buildkite check.



> Thanks @mmahadevan108 !
> 
> Please add a board document. See `boards/arm/lpcxpresso55s69/doc/index.rst` and `boards/arm/mimxrt1050_evk/doc/index.rst` for example.
> 
> Please also fix the errors flagged in the failed compliance check. See https://github.com/zephyrproject-rtos/zephyr/pull/26785/checks?check_run_id=856037114
> 
> Moving the nxp hal update commit into this PR will help fix the failed buildkite check.

I am confused by this ERROR flagged by the compliance check:

:416: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
416: FILE: drivers/gpio/gpio_mcux_lpc.c:42:
+	IOPCTL_Type *pinmux_base;

I am following the same convention as it done in the rest of the file, what should I be using?

MaureenHelm commented:
> I am confused by this ERROR flagged by the compliance check:
> 
> :416: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
> 416: FILE: drivers/gpio/gpio_mcux_lpc.c:42:
> 
> * IOPCTL_Type *pinmux_base;
> 
> I am following the same convention as it done in the rest of the file, what should I be using?

This looks ok to me. Maybe it's an issue with the checkpatch script? @nashif can you have a look?

agansari commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.yaml:
```diff
@@ -0,0 +1,20 @@
+#
```
File name does not need '_cm33' suffix.
Other boards add this suffix because they are multi-core and have asymmetric architecture, e.g. M4+M0.

mmahadevan108 replied:
This was added to have the flexibility to add zephyr support for the DSP core at a later time.

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/linker.ld:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2014 Wind River Systems, Inc.
```
Thank you for catching this. Its a copy/paste error. I have changed this to the NXP copyright

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,207 @@
+		CLOCK_EnableSysOscClk(true, true, BOARD_SYSOSC_SETTLING_US); /* Enable system OSC */
+	}
+	/* Updated XTAL oscillator settling time */
+	/* POWER_UpdateOscSettlingTime(BOARD_SYSOSC_SETTLING_US); */
```
Done, I thought of leaving this in so we can add it back when the power library becomes available. However I do have this change as part of a separate commit, so we could simply undo the commit to raise the core frequency.

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,207 @@
+	CLOCK_AttachClk(kMAIN_PLL_to_MAIN_CLK);
+
+	/* Configure Audio PLL clock source */
+	CLOCK_InitAudioPll(&g_audioPllConfig);
```
Done, I have added configs called INIT_SYS_PLL and INIT_AUDIO_PLL similar to what is done for other RT platforms

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,207 @@
+	.sys_pll_mult = kCLOCK_SysPllMult22  /* Divide by 22 */
+};
+
+const clock_audio_pll_config_t g_audioPllConfig = {
```
done

mmahadevan108 commented:
> Thanks @mmahadevan108 !
> 
> Please add a board document. See `boards/arm/lpcxpresso55s69/doc/index.rst` and `boards/arm/mimxrt1050_evk/doc/index.rst` for example.
> 
> Please also fix the errors flagged in the failed compliance check. See https://github.com/zephyrproject-rtos/zephyr/pull/26785/checks?check_run_id=856037114
> 
> Moving the nxp hal update commit into this PR will help fix the failed buildkite check.

I have added the documentation for RT600 and moved the hal update commit into this PR. The only thing I am not clear on is the sanity check failures.

mmahadevan108 committed:
drivers: gpio: Init the PINT module

Need to call PINT_Init to initialize the block
before using it. This fixes the issue seen on
RT600 with the button example

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

agansari commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -296,6 +296,7 @@ static int gpio_mcux_lpc_pin_interrupt_configure(struct device *dev,
 		data->pint_id[pin] = pint_idx;
 	}
 
+	PINT_Init(config->pint_base);
```
Move this to soc.c like lpc55s69

mmahadevan108 replied:
Since PINT is used only in this file, wouldn't this be a place to put this instead of soc.c?

agansari replied:
@mmahadevan108 PINT initialization is done at soc level as it's serving multiple gpio device 0 and 1, you may end up initializing it twice, with no side effect; also it's similar to clock setup done in soc.c, clocks setup and hw block initialization serving other devices.

mmahadevan108 replied:
This call is not SoC specific like the clocking setup. I would be duplicating the same line in all soc files, also it leaves the chance for future soc additions to miss calling this function.
I can add a flag check so this gets initialized once.

MaureenHelm commented on soc/arm/nxp_imx/rt6xx/Kconfig.soc:
```diff
@@ -0,0 +1,89 @@
+	  choice defines the default value for this string.
+
+config INIT_SYS_PLL
+        bool "Initialize SYS PLL"
```
Replace the leading spaces with a tab

mmahadevan108 replied:
Done

MaureenHelm commented on soc/arm/nxp_imx/rt6xx/Kconfig.soc:
```diff
@@ -0,0 +1,89 @@
+        bool "Initialize SYS PLL"
+
+config INIT_AUDIO_PLL
+        bool "Initialize Audio PLL"
```
Replace the leading spaces with a tab

mmahadevan108 replied:
Done

MaureenHelm commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,216 @@
+	}
+	/* Updated XTAL oscillator settling time */
+	POWER_UpdateOscSettlingTime(BOARD_SYSOSC_SETTLING_US);
+	CLOCK_SetXtalFreq(BOARD_XTAL_SYS_CLK_HZ);			   /* Sets external XTAL OSC freq */
```
Please remove the trailing comments in this section. They cause a lot of checkpatch warnings and don't provide much information.

mmahadevan108 committed:
MXRT600: Lower clock speed to avoid using power library

We cannot call into the power library API as it is currently
available in binary format which cannot be included

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

MaureenHelm reviewed (changes requested):
Please have a look at the buildkite failure. The `mm_swiftio` board (rt1050) is missing the `dcd_data` structure due to the change in `modules/hal/nxp/mcux/boards/CMakeLists.txt`:
```cmake
if (${MCUX_BOARD} MATCHES "evk[bm]imxrt1[0-9][0-9][0-9]")
zephyr_compile_definitions_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER XIP_BOOT_HEADER_ENABLE=1)                                                                                                                                                      
zephyr_compile_definitions_ifdef(CONFIG_DEVICE_CONFIGURATION_DATA XIP_BOOT_HEADER_DCD_ENABLE=1)
zephyr_library_sources_ifdef(CONFIG_BOOT_FLEXSPI_NOR ${MCUX_BOARD}/${MCUX_BOARD}_flexspi_nor_config.c)
zephyr_library_sources_ifdef(CONFIG_DEVICE_CONFIGURATION_DATA ${MCUX_BOARD}/${MCUX_BOARD}_sdram_ini_dcd.c)
elseif (${MCUX_BOARD} MATCHES "evkmimxrt6[0-9][0-9]")
zephyr_compile_definitions_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER BOOT_HEADER_ENABLE=1)
zephyr_library_sources_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER ${MCUX_BOARD}/flash_config.c)
endif()
```

By moving `XIP_BOOT_HEADER_DCD_ENABLE=1` inside the evk board conditional, we're not setting it anymore for the `mm_swiftio` board.

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,216 @@
+	.audio_pll_src	= kCLOCK_AudioPllXtalIn, /* OSC clock */
+	.numerator		= 5040,					 /* Numerator of the SYSPLL0 fractional loop divider isnull */
+	.denominator	= 27000,				 /* Denominator of the SYSPLL0 fractional loop divider isnull */
+	.audio_pll_mult = kCLOCK_AudioPllMult22  /* Divide by 22 */
```
Done

mmahadevan108 commented on soc/arm/nxp_imx/rt6xx/soc.c:
```diff
@@ -0,0 +1,216 @@
+	z_arm_pendsv,							/* 0x38 */
+#if defined(CONFIG_SYS_CLOCK_EXISTS)
+	z_clock_isr,							/* 0x3C */
+#else
```
Done

mmahadevan108 commented:
> Please have a look at the buildkite failure. The `mm_swiftio` board (rt1050) is missing the `dcd_data` structure due to the change in `modules/hal/nxp/mcux/boards/CMakeLists.txt`:
> 
> ```cmake
> if (${MCUX_BOARD} MATCHES "evk[bm]imxrt1[0-9][0-9][0-9]")
> zephyr_compile_definitions_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER XIP_BOOT_HEADER_ENABLE=1)                                                                                                                                                      
> zephyr_compile_definitions_ifdef(CONFIG_DEVICE_CONFIGURATION_DATA XIP_BOOT_HEADER_DCD_ENABLE=1)
> zephyr_library_sources_ifdef(CONFIG_BOOT_FLEXSPI_NOR ${MCUX_BOARD}/${MCUX_BOARD}_flexspi_nor_config.c)
> zephyr_library_sources_ifdef(CONFIG_DEVICE_CONFIGURATION_DATA ${MCUX_BOARD}/${MCUX_BOARD}_sdram_ini_dcd.c)
> elseif (${MCUX_BOARD} MATCHES "evkmimxrt6[0-9][0-9]")
> zephyr_compile_definitions_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER BOOT_HEADER_ENABLE=1)
> zephyr_library_sources_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER ${MCUX_BOARD}/flash_config.c)
> endif()
> ```
> 
> By moving `XIP_BOOT_HEADER_DCD_ENABLE=1` inside the evk board conditional, we're not setting it anymore for the `mm_swiftio` board.

I have changed this implementation to add a new config option called NXP_IMX_RT6XX_BOOT_HEADER. This should help fix this issue.

mmahadevan108 commented:
Anything else needed for this PR?

pabigot commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -236,6 +252,7 @@ static int gpio_mcux_lpc_pin_interrupt_configure(struct device *dev,
 	uint32_t port = config->port_no;
 	uint32_t isr;
 	uint32_t pint_idx;
+	static bool pint_inited;
```
Is that a MISRA requirement?  C guarantees it'll be zero-initialized (false).

mmahadevan108 replied:
static variables are initialized to 0.

dleach02 replied:
Not sure. Maybe it is a personal preference on my part that I prefer initialization. If it is guaranteed then we don't need to do it.

mmahadevan108 commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -0,0 +1,98 @@
+&red_led {
+	status = "okay";
+};
+
```
Done

MaureenHelm commented on west.yml:
```diff
@@ -98,7 +98,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 478f03a71256aeb2890e685ebf5a1dc6832a6b5f
+      revision: pull/49/head
```
I merged the hal PR, so please update this to the new SHA.

mmahadevan108 replied:
Done

MaureenHelm reviewed (changes requested):
Almost there. Thanks @mmahadevan108

MaureenHelm commented:
Overriding checkpatch false positive, based on #27035
